### PR TITLE
refs #1947 added broken-logic-precedence-warn parse directive and all…

### PR DESCRIPTION
**Pull request by pavelkveton**
… the infrastructure

### CODE REVIEW

File: doxygen/lang/245_parse_directives.dox.tmpl [hunk 1]
```diff
@@ -262,18 +262,47 @@
     Use old pre-0.8.12 precedence where the \ref logical_and_operator "&&" had the same precedence as \ref logical_or_operator "||" and all three bitwise operators (
     \ref bitwise_and_operator "&", \ref bitwise_or_operator "|", \ref bitwise_xor_operator "^") shared the same precedence.
 
+    An example of impact of @ref broken-logic-precedence "%broken-logic-precedence" directive:
```
Comment (davidnich) [inline doxygen/lang/245_parse_directives.dox.tmpl:265]: this does not render correctly - you can use a table, but not in this format - you need to remove the -----s for example

File: doxygen/lang/245_parse_directives.dox.tmpl [hunk 2]
```diff
@@ -262,18 +262,47 @@
...
+
+    @par Parse Option Constant:
+    @ref Qore::PO_BROKEN_LOGIC_PRECEDENCE_WARN
+
+    @par Description:
+    Can only be set with @ref broken-logic-precedence "%broken-logic-precedence". When set on, a parse-time warning is printed for expressions that are affected by @ref broken-logic-precedence "%broken-logic-precedence".
+
+    @see @ref broken-logic-precedence "%broken-logic-precedence"
+
+    @since %Qore 0.8.12
```
Comment (davidnich) [inline doxygen/lang/245_parse_directives.dox.tmpl]: the version should be 0.8.12.11

File: doxygen/lang/900_release_notes.dox.tmpl [hunk 3]
```diff
@@ -301,6 +301,7 @@
       - @ref Qore::ParseOptionStringMap
       - @ref Qore::PO_BROKEN_LIST_PARSING
       - @ref Qore::PO_BROKEN_LOGIC_PRECEDENCE
+      - @ref Qore::PO_BROKEN_LOGIC_PRECEDENCE_WARN
```
Comment (davidnich) [inline doxygen/lang/900_release_notes.dox.tmpl]: this needs to be in the new release notes for 0.8.12.11, we shouldn't update release notes for releases already out (for a long time :))

File: lib/QC_Program.qpp [hunk 4]
```diff
@@ -419,6 +419,13 @@ const PO_BROKEN_LIST_PARSING = PO_BROKEN_LIST_PARSING;
 */
 const PO_BROKEN_LOGIC_PRECEDENCE = PO_BROKEN_LOGIC_PRECEDENCE;
 
+//! allows for warn before expressions affected by @ref broken-logic-precedence "%broken-logic-precedence"
+/** @see @ref broken-logic-precedence-warn "%broken-logic-precedence-warn"
+
+    @since %Qore 0.8.12
```
Comment (davidnich) [inline lib/QC_Program.qpp]: incorrect version as above

File: lib/QoreProgram.cpp [hunk 5]
```diff
@@ -165,6 +166,18 @@ void qore_program_private_base::setDefines() {
    }
 }
 
+void qore_program_private_base::checkDefines(ExceptionSink *xsink) {
+    dmap_t::iterator blp = dmap.find("PO_BROKEN_LOGIC_PRECEDENCE");
+    dmap_t::iterator blpw = dmap.find("PO_BROKEN_LOGIC_PRECEDENCE_WARN");
+    if (blp == dmap.end() || !blp->second || !blp->second->getAsBool()) {
+        if (blpw != dmap.end() && blpw->second && blpw->second->getAsBool()) {
+            if (xsink) {
+                xsink->raiseException("PARSE-DIRECTIVE-ERROR", "broken-logic-precedence-warn cannot be set without broken-logic-precedence");
```
Comment (davidnich) [inline lib/QoreProgram.cpp]: this actually should not be a parse option but rather a warning, which would be a different implementation, but would make more sense

**Review by davidnich — CHANGES_REQUESTED** [review]
I believe that instead of a new parse option that this should be set as a warning that takes effect only when %broken-logic-precedence is set, because we already have a solution for warnings - let's please discuss if necessary.

File: doxygen/lang/245_parse_directives.dox.tmpl (hunk 1 above)
Comment (pavelkveton) [reply doxygen/lang/245_parse_directives.dox.tmpl:265]: fixed in the latest commit

File: doxygen/lang/245_parse_directives.dox.tmpl (hunk 2 above)
Comment (pavelkveton) [reply doxygen/lang/245_parse_directives.dox.tmpl]: fixed in the latest commit

File: doxygen/lang/900_release_notes.dox.tmpl (hunk 3 above)
Comment (pavelkveton) [reply doxygen/lang/900_release_notes.dox.tmpl]: with the latest commit this line vanishes

File: lib/QC_Program.qpp (hunk 4 above)
Comment (pavelkveton) [reply lib/QC_Program.qpp]: fixed in the latest commit

File: lib/QoreProgram.cpp (hunk 5 above)
Comment (pavelkveton) [reply lib/QoreProgram.cpp]: yes, sure - fixed

File: qore.spec-fedora [hunk 6]
```diff
@@ -165,6 +165,9 @@ make check
 %{_mandir}/man1/qore.1.*
 
 %changelog
+* Thu Apr 13 2017 David Nichols <david@qore.org> 0.8.12.11-1
```
Comment (davidnich) [inline qore.spec-fedora]: please fix the date here - also you can put your name & email if you want :)

**Comment by davidnich** [comment]
@pavelkveton it looks good now except one very minor change to a spec file

**Review by davidnich — APPROVED** [review]

**Pull request merged**